modules/pricer/src/main/java/com/opengamma/strata/pricer/impl/volatility/smile/function/SABRHaganVolatilityFunctionProvider.java

```diff
@@ -0,0 +1,993 @@
+        // so we return an arbitrary large number
+        volatilityAdjoint[3] = 1e7;
+      }
+      return volatilityAdjoint;
+    }
+
+    // Implementation note: Forward sweep.
+    double sfK = Math.pow(forward * strike, betaStar / 2);
+    double lnrfK = Math.log(forward / strike);
+    double z = nu / alpha * sfK * lnrfK;
+    double sf1 = sfK *
+        (1 + betaStar * betaStar / 24 * (lnrfK * lnrfK) + Math.pow(betaStar, 4) / 1920 * Math.pow(lnrfK, 4));
+    double sf2 = (1 + (Math.pow(betaStar * alpha / sfK, 2) / 24 + (rho * beta * nu * alpha) / (4 * sfK) + (2 - 3 *
+        rho * rho) *
+        nu * nu / 24) *
+        timeToExpiry);
+
+    // Implementation note: Backward sweep.
+    double[] zOverChi = zOverChiWithDev(rho, z);
+    double vBar = 1;
+    double sf2Bar = alpha / sf1 * zOverChi[0] * vBar;
+    double sf1Bar = -alpha / (sf1 * sf1) * zOverChi[0] * sf2 * vBar;
+    double rzxzBar = alpha / sf1 * sf2 * vBar;
+    double zBar = zOverChi[2] * rzxzBar;
+    //    double xzBar = 0;
+
+    double sfKBar = nu / alpha * lnrfK * zBar + sf1 / sfK * sf1Bar -
+        (Math.pow(betaStar * alpha, 2) / Math.pow(sfK, 3) / 12 + (rho * beta * nu * alpha) / 4 / (sfK * sfK)) *
+        timeToExpiry
+        * sf2Bar;
+
+    double nuBar = 1 / alpha * sfK * lnrfK * zBar +
+        ((rho * beta * alpha) / (4 * sfK) + (2 - 3 * rho * rho) * nu / 12) * timeToExpiry * sf2Bar;
+    double rhoBar = zOverChi[1] * rzxzBar + ((beta * nu * alpha) / (4 * sfK) - rho * nu * nu / 4) * timeToExpiry *
+        sf2Bar;
+
+    double alphaBar = -nu / (alpha * alpha) * sfK * lnrfK * zBar +
+        ((betaStar * alpha / sfK) * (betaStar / sfK) / 12 + (rho * beta * nu) / (4 * sfK)) * timeToExpiry * sf2Bar + 1 /
+        sf1
+        * zOverChi[0] * sf2 * vBar;
+    double betaBar = -0.5 * Math.log(forward * strike) * sfK * sfKBar - sfK *
+        (betaStar / 12 * (lnrfK * lnrfK) + Math.pow(betaStar, 3) / 480 * Math.pow(lnrfK, 4)) * sf1Bar
+        + (-betaStar * alpha * alpha / sfK / sfK / 12 + rho * nu * alpha / 4 / sfK) * timeToExpiry * sf2Bar;
+
+    volatilityAdjoint[0] = alphaBar;
+    volatilityAdjoint[1] = betaBar;
+    volatilityAdjoint[2] = rhoBar;
+    volatilityAdjoint[3] = nuBar;
+
+    return volatilityAdjoint;
+  }
+
+  /**
+   * Computes the Black implied volatility in the SABR model and its derivatives.
+   * <p>
+   * This returns an array with [0] the volatility, [1] Derivative w.r.t the forward, [2] the derivative w.r.t the strike, 
+   * [3] the derivative w.r.t. to alpha, [4] the derivative w.r.t. to beta, [5] the derivative w.r.t. to rho, and 
+   * [6] the derivative w.r.t. to nu.
+   * 
+   * @param option The option.
+   * @param forward The forward value of the underlying
+   * @param data The SABR data.
+   * @return the volatility and sensitivities
+   */
+  public double[] getVolatilityAdjoint(EuropeanVanillaOption option, double forward, SABRFormulaData data) {
```

Comment:
For "Adjoint AD" version of functions, we should standardize the type of object which is returned. I will create a "DoubleDerivatives" or "ValueDerivatives" object as a return type.